File: Python/DataStructures/SequentialDeque.py
from utils.utils import emptyDequeException

class SequentialDeque():
    def __init__(self):

        self.elements = []
        self.front = 0
        self.rear = -1

    def size(self):
        return self.rear + 1
    
    def isEmpty(self):
        return self.size() == 0

    def push(self, e):
        self.rear += 1
        self.elements.insert(0, e)
    
    def pop(self):
        if self.isEmpty():
            return emptyDequeException
        e = self.elements.pop(0)
        self.rear -= 1
        return e
    
    def inject(self, e):
        self.rear += 1
        self.elements.append(e)
    
    def eject(self):
        if self.isEmpty():
            return emptyDequeException
        e = self.elements.pop(self.rear)
        self.rear -= 1
        return e

File: Golang/DataStructures/SequentialDeque.go
package DataStructures

import "github.com/JhyeonLee/DataStructuresAndAlgorithms/golang/utils"

type SequentialDeque struct {
	Elements []int
	front    int
	rear     int
}

func Initialize_SequentialDeque() *SequentialDeque {
	deque := &SequentialDeque{
		front: 0,
		rear:  -1,
	}
	return deque
}

func (Dq *SequentialDeque) Size() int {
	return Dq.rear + 1
}

func (Dq *SequentialDeque) IsEmpty() bool {
	return Dq.Size() == 0
}

func (Dq *SequentialDeque) Push(e int) {
	// fullQueueException()

	Dq.rear++
	Dq.Elements = append([]int{e}, Dq.Elements...)
}

func (Dq *SequentialDeque) Pop() (int, error) {
	if Dq.IsEmpty() {
		return -1, utils.EmptyDequeException()
	}
	// fullQueueException()
	e := Dq.Elements[Dq.front]
	Dq.rear--
	Dq.Elements = Dq.Elements[Dq.front+1:]
	return e, utils.EmptyDequeException()
}

func (Dq *SequentialDeque) Inject(e int) {
	// fullQueueException()

	Dq.rear++
	Dq.Elements = append(Dq.Elements, e)
}

func (Dq *SequentialDeque) Eject() (int, error) {
	if Dq.IsEmpty() {
		return -1, utils.EmptyDequeException()
	}
	// fullQueueException()
	e := Dq.Elements[Dq.rear]
	Dq.rear--
	Dq.Elements = Dq.Elements[:Dq.rear+1]
	return e, utils.EmptyDequeException()
}
